main/main.go

```diff
@@ -254,76 +265,178 @@ func main() {
 	)
 }
 
-// runRelayer creates a relayer instance for a subnet. It listens for warp messages on that subnet, and handles delivery to the destination
-func runRelayer(
+// runListener creates a Listener instance and the ApplicationRelayers for a subnet.
+// The Listener listens for warp messages on that subnet, and the ApplicationRelayers handle delivery to the destination
+func runListener(
 	ctx context.Context,
 	logger logging.Logger,
 	metrics *relayer.ApplicationRelayerMetrics,
 	db database.RelayerDatabase,
 	ticker *utils.Ticker,
-	sourceSubnetInfo config.SourceBlockchain,
+	sourceBlockchain config.SourceBlockchain,
 	network *peers.AppRequestNetwork,
-	responseChan chan message.InboundMessage,
 	destinationClients map[ids.ID]vms.DestinationClient,
 	messageCreator message.Creator,
 	relayerHealth *atomic.Bool,
-	manualWarpMessages []*relayerTypes.WarpLogInfo,
+	manualWarpMessages []*relayerTypes.WarpMessageInfo,
 	cfg *config.Config,
 ) error {
-	logger.Info(
-		"Creating relayer",
-		zap.String("originBlockchainID", sourceSubnetInfo.BlockchainID),
+	// Dial the eth client
+	ethClient, err := ethclient.DialWithConfig(
+		context.Background(),
+		sourceBlockchain.RPCEndpoint.BaseURL,
+		sourceBlockchain.RPCEndpoint.HTTPHeaders,
+		sourceBlockchain.RPCEndpoint.QueryParams,
 	)
+	if err != nil {
+		logger.Error(
+			"Failed to connect to node via RPC",
+			zap.String("blockchainID", sourceBlockchain.BlockchainID),
+			zap.Error(err),
+		)
+		return err
+	}
 
-	listener, err := relayer.NewListener(
+	// Create the ApplicationRelayers
+	applicationRelayers, minHeight, err := createApplicationRelayers(
+		ctx,
 		logger,
 		metrics,
 		db,
 		ticker,
-		sourceSubnetInfo,
+		sourceBlockchain,
 		network,
-		responseChan,
-		destinationClients,
 		messageCreator,
+		cfg,
+		ethClient,
+		destinationClients,
+	)
+	if err != nil {
+		logger.Error(
+			"Failed to create application relayers",
+			zap.String("blockchainID", sourceBlockchain.BlockchainID),
+			zap.Error(err),
+		)
+		return err
+	}
+	logger.Info(
+		"Created application relayers",
+		zap.String("blockchainID", sourceBlockchain.BlockchainID),
+	)
+
+	// Create the Listener
+	listener, err := relayer.NewListener(
+		logger,
+		sourceBlockchain,
 		relayerHealth,
 		cfg,
+		applicationRelayers,
+		minHeight,
+		ethClient,
 	)
 	if err != nil {
 		return fmt.Errorf("failed to create listener instance: %w", err)
 	}
 	logger.Info(
 		"Created listener",
-		zap.String("blockchainID", sourceSubnetInfo.BlockchainID),
+		zap.String("blockchainID", sourceBlockchain.BlockchainID),
 	)
-
-	// Send any messages that were specified in the configuration
-	for _, warpMessage := range manualWarpMessages {
-		logger.Info(
-			"Relaying manual Warp message",
-			zap.String("blockchainID", sourceSubnetInfo.BlockchainID),
-			zap.String("warpMessageBytes", hex.EncodeToString(warpMessage.UnsignedMsgBytes)),
+	err = listener.ProcessManualWarpMessages(logger, manualWarpMessages, sourceBlockchain)
```

Comment:
I think the structure of the initialization flow here would be more readable and extendable if it roughly broke down as:
- Create the necessary application relayer instances.
- Create the necessary listener instances with the application relayers.
- Start `ProcessLogs` routines for each listener instance.
- Process the manual warp messages with the respective listener/application relayers. (I think it's a bit odd that `ProcessManualWarpMessages` is part of the `listener` since it isn't actually related to listening for events on that chain.)